WP Stories integration - save event snackbars

mzorz opened the pull request:
This PR builds on top of #11930 

This PR implements snackbars on the MySiteFragment for:
- start Story saving process
- Story saving process ending successfully
- Story saving process ending in error

This PR uses a similar structure for event posting with EventBus as per the Media and Post uploaders in the `UploadService`, with some analogies for example the `StorySaveEvents` class holds all possible events for the Saving process now (similarly to how `PostEvents` holds events for Post uploading), and hence we can now listen to these in a more structured way.

1. moved start and results events into its own separate class so it's easier to read (`StorySaveEvents`)
2. added the EventBus listeners to MySite fragment as per UploadErrorEvent
3. to handle snackbar through the sequencer, we're using the  `UploadUtilsWrapper` which was already there serving the same purpose. Given both the `FrameSaveService` and `UploadService` are going to be tightly connected (one chaining into the other), makes sense to reuse the same Utils class for snackbars instead of duplicating code unnecessarily.


To test:

Successful path
1. tap on the FAB
2. select Story
3. when the camera preview appears, take a picture
4. optionally, add emoji  / text
5. tap on the `PUBLISH` buttton on the top-right corner of the screen
6. verify the app returns to the previous (main) screen and both he start and end snackbars are shown.

Error path
1. introduce an error in saving (i.e. artificial exception in mp4composerengine such as the one int [this gist](https://gist.github.com/mzorz/da69f0dea9918877defefeb0573b560d))
2. create a video and add some text or emoji
3. tap publish
4. observe the error notification and error snackbar appear
5. tap on MANAGE action on the error snackbar
6. observe the error handling view appears
7. tap RETRY,   see it still fails
8. tap on the Cross to exit the screen, and when prompted about discarding the Story, tap DISCARD
9. observe no snackbar appears (meaning, the sticky event added by the Service has been successfully removed - see commit 5397a25fbeba391f1d795aba429129300d8391fe in Stories repo)
10. observe the error notification (originally created and appearing in step 4 above) is not there anymore in the system dashboard (meaning it has been correctly cleard when discarding the Sotry on step 8 above). (see related commit bbce9a03e53b7c8cb0da3ded5b8da65bdabfd1da in Stories repo, and de7d5c50b2b6e68d203f048c987cabf1ca36cd73 in WPAndroid repo)


PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

jd-alexander commented on WordPress/src/main/java/org/wordpress/android/ui/main/MySiteFragment.kt:
```diff
@@ -1051,6 +1060,73 @@ class MySiteFragment : Fragment(),
+                        // errored story but the error notification will remain existing in the system dashboard)
+                        intent.action = getNotificationIdForError(event.storyIndex).toString() + ""
+
+                        // TODO add NotificationType.MEDIA_SAVE_ERROR param later when integrating with WPAndroid
+                        //        val notificationType = NotificationType.MEDIA_SAVE_ERROR
+                        //        notificationIntent.putExtra(ARG_NOTIFICATION_TYPE, notificationType)
```
Do we want to do this integration now? or it's intentional left for a later time. Let me know. 🙏

mzorz replied:
Good question! Yes we need to do that as part of the integration, but it's tackled on a later PR https://github.com/wordpress-mobile/WordPress-Android/pull/12074 👍  (see this [comment](https://github.com/wordpress-mobile/WordPress-Android/pull/12074/files#r434053979) in particular)

jd-alexander replied:
Oh nice! Thanks!

jd-alexander commented on WordPress/src/main/java/org/wordpress/android/ui/main/MySiteFragment.kt:
```diff
@@ -1051,6 +1060,73 @@ class MySiteFragment : Fragment(),
+    fun onEventMainThread(event: StorySaveResult) {
+        EventBus.getDefault().removeStickyEvent(event)
+        if (event.isSuccess()) {
+            // TODO WPSTORIES add TRACKS
+            // AnalyticsTracker.track(Stat.MY_SITE_ICON_UPLOAD_UNSUCCESSFUL);
```
This is a nitpick 😅 when I read this, I got a bit confused because I had to process that the event in `track` wasn't related to this code, which is expected since the `TODO` is above. So could we do without the `AnalyticsTracker` here? Also, I think we have a wrapper around the `AnalyticsTracker` as well to make classes more testable.

mzorz replied:
:D I know what you mean - and this is all tackled in https://github.com/wordpress-mobile/WordPress-Android/pull/12074 👍

jd-alexander replied:
Okay understood! As I said, I know I will comment on a few things that are fixed in future PRs, so bear with me 🙇

jd-alexander commented on WordPress/src/main/java/org/wordpress/android/ui/main/MySiteFragment.kt:
```diff
@@ -1051,6 +1060,73 @@ class MySiteFragment : Fragment(),
+                    snackbarMessage
+            )
+        } else {
+            // TODO WPSTORIES add TRACKS
+            // AnalyticsTracker.track(Stat.MY_SITE_ICON_UPLOAD_UNSUCCESSFUL);
```
Same thought as above.

mzorz replied:
sure, refer to https://github.com/wordpress-mobile/WordPress-Android/pull/12074 👍

jd-alexander commented on WordPress/src/main/java/org/wordpress/android/ui/main/MySiteFragment.kt:
```diff
@@ -1051,6 +1060,73 @@ class MySiteFragment : Fragment(),
 
+    @Subscribe(sticky = true, threadMode = ThreadMode.MAIN)
+    fun onEventMainThread(event: StorySaveResult) {
+        EventBus.getDefault().removeStickyEvent(event)
```
Another nitpick, I think we can leave this, but since we are trying to lessen the usage of these `static` calls directly, we can use the `EventBusWrapper`. Wdyt?

mzorz replied:
given this is a Fragment I don't think there's much of a difference, but I may be missing something. That said, we can definitely give it another round for alpha2 when we change focus more on testability, sounds good?

jd-alexander replied:
Yes, sounds good. Yes, this fragment won't be tested so we can leave this as is and look at it later. Thanks for the comment 🙏

jd-alexander commented:
Hey, @mzorz I just realized that when I have to enable to microphone permissions, the capture circle doesn't get shown for some reason.

<kbd><img src="https://user-images.githubusercontent.com/1509205/84970021-7a0ba300-b0df-11ea-9117-c50d857f832e.gif" width="320"></kbd>

jd-alexander commented:
While testing the successful path on my Pixel 4 XL, I tried to add emoji to the frame but then I got this crash/ ANR. Let me know what you think 🙏 

``` Process: org.wordpress.android.beta, PID: 14819
    java.lang.IllegalStateException: EmojiCompat is not initialized. Please call EmojiCompat.init() first
        at androidx.core.util.Preconditions.checkState(Preconditions.java:151)
        at androidx.emoji.text.EmojiCompat.get(EmojiCompat.java:357)
        at com.wordpress.stories.compose.emoji.EmojiPickerFragment$EmojiAdapter.onBindViewHolder(EmojiPickerFragment.kt:108)
        at com.wordpress.stories.compose.emoji.EmojiPickerFragment$EmojiAdapter.onBindViewHolder(EmojiPickerFragment.kt:93)
        at androidx.recyclerview.widget.RecyclerView$Adapter.onBindViewHolder(RecyclerView.java:7065)
        at androidx.recyclerview.widget.RecyclerView$Adapter.bindViewHolder(RecyclerView.java:7107)
        at androidx.recyclerview.widget.RecyclerView$Recycler.tryBindViewHolderByDeadline(RecyclerView.java:6012)
        at androidx.recyclerview.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:6279)
        at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6118)
        at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6114)
        at androidx.recyclerview.widget.LinearLayoutManager$LayoutState.next(LinearLayoutManager.java:2303)
        at androidx.recyclerview.widget.GridLayoutManager.layoutChunk(GridLayoutManager.java:561)
        at androidx.recyclerview.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1587)
        at androidx.recyclerview.widget.LinearLayoutManager.onLayoutChildren(LinearLayoutManager.java:665)
        at androidx.recyclerview.widget.GridLayoutManager.onLayoutChildren(GridLayoutManager.java:170)
        at androidx.recyclerview.widget.RecyclerView.dispatchLayoutStep2(RecyclerView.java:4134)
        at androidx.recyclerview.widget.RecyclerView.onMeasure(RecyclerView.java:3540)
        at android.view.View.measure(View.java:25086)```

jd-alexander commented:
Without adding emoji I was able to verify that the behavior below works 😄 

```
Successful path

tap on the FAB
select Story
when the camera preview appears, take a picture
optionally, add emoji / text
tap on the PUBLISH buttton on the top-right corner of the screen
verify the app returns to the previous (main) screen and both he start and end snackbars are shown.
```

mzorz commented:
Thanks for the review @jd-alexander !
1. Camera capture button not present right after permissions request (first time) is already tracked in our [project](https://github.com/Automattic/portkey-android/issues/346).
2. The EmojiCompat exception mentioned [above](https://github.com/wordpress-mobile/WordPress-Android/pull/11931#issuecomment-645731021)  is fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/12045

> Without adding emoji I was able to verify that the success and error path instructions worked well and produced the expected results. Once we have addressed the issues above we are ready to go 😄

I guess this one should be ready to go then 😎 👍

mzorz commented:
> Once we have addressed the issues above we are ready to go 😄

Given these are tracked as explained above, and now that the rest of the WP PRs cascade is updated, merging this one 👍
